src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs

```diff
@@ -19,11 +19,34 @@ namespace Microsoft.CodeAnalysis.CodeRefactorings
     {
         /// <summary>
         /// Document corresponding to the <see cref="CodeRefactoringContext.Span"/> to refactor.
+        /// For code refactorings that support non-source documents by providing a non-default value for
+        /// <see cref="ExportCodeRefactoringProviderAttribute.DocumentKinds"/>, this property will
+        /// throw an <see cref="InvalidOperationException"/>. Such refactorings should use the
+        /// <see cref="CodeRefactoringContext.TextDocument"/> property instead.
         /// </summary>
-        public Document Document { get; }
+        public Document Document
+        {
+            get
+            {
+                if (TextDocument is not Document document)
+                {
+                    throw new InvalidOperationException();
```

Comment:
Should we put a string here to make it clearer what the mistake was?

Reply:
Sure, let me add an exception message.